dataServices.js

```diff
@@ -0,0 +1,59 @@
+'use strict';
+
+const products = require('./src/data/products');
+const categories = require('./src/data/categories');
+
+let productMap = null;
+let categoryMap = null;
+let combinedProductMap = null;
+
+// Returns a map of all products, with their ids as keys and whole objects as values.
+function getProducts() {
+  if (!productMap) {
+    let newProductMap = {};
+    products.products.forEach(product => {
+      newProductMap[product.id] = product;
+    });
+    productMap = newProductMap;
+  }
+  return productMap;
+}
+
+// Returns a map of all categories, with their ids as keys and whole objects as values.
+function getCategories() {
+  if (!categoryMap) {
+    let newCategoryMap = {};
+    categories.categories.forEach(category => {
+      newCategoryMap[category.id] = category;
+    });
+    categoryMap = newCategoryMap;
+  }
+  return categoryMap;
+}
+
+// Returns a map of all products with category names injected, with the product ids as keys and product objects as values.
+function getCombinedProductMap() {
+  if (!combinedProductMap) {
+    combineProductsWithCategories();
+  }
+  return combinedProductMap;
+}
+
+// Injects category names into the product objects
+function combineProductsWithCategories() {
+  let combinedMap = getProducts();
+  let categoryMap = getCategories();
+  Object.keys(combinedMap).forEach(key => {
```

Comment:
Use [`.map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) instead, operating on the `combinedMap` and adding the `categoryName` attribute while mapping it.

Reply:
Thanks for the tip, learned something new!

Reply:
After testing I realised that combinedMap is an object (not an array) and doesn't have the function `map()`... oops!